### PR TITLE
Utilize new APIs in the WebSocket implementation

### PR DESCRIPTION
**Pull request by PaulusParssinen**
Use few newer APIs in the managed websocket implementation. Gets rid of little manual endianness bit-twiddling that is common for code before `BinaryPrimitives`.

### CODE REVIEW

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs [hunk 1]
```diff
@@ -210,12 +211,12 @@ private void ReleaseBuffer()
...
 
-        private static unsafe bool IsFinished(ZLibStreamHandle stream, out byte? remainingByte)
+        private static bool IsFinished(ZLibStreamHandle stream, out byte? remainingByte)
         {
             // There is no other way to make sure that we've consumed all data
             // but to try to inflate again with at least one byte of output buffer.
-            byte b;
-            if (Inflate(stream, new Span<byte>(&b, 1), FlushCode.SyncFlush) == 0)
+            byte b = 0;
+            if (Inflate(stream, MemoryMarshal.CreateSpan(ref b, 1), FlushCode.SyncFlush) == 0)
```
Comment (stephentoub) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs:219]: This should just be `new Span(ref b) `

Comment (PaulusParssinen) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs]: Right, Thanks! https://github.com/dotnet/runtime/pull/101953/commits/4708726602b0a364583010d4914440840531da23

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 2]
```diff
@@ -1495,9 +1490,7 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
 
                 if (toMaskEnd - toMaskPtr >= sizeof(int))
                 {
-                    int rolledMask = BitConverter.IsLittleEndian ?
-                        (int)BitOperations.RotateRight((uint)mask, maskIndex * 8) :
-                        (int)BitOperations.RotateLeft((uint)mask, maskIndex * 8);
+                    int rolledMask = (int)BitOperations.RotateRight((uint)mask, maskIndex * 8);
```
Comment (PaulusParssinen) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:1493]: I was thinking of using `TensorPrimitives.Xor` here for the `sizeof(int)` elements but I think we can't guarantee alignment of the `toMask` for `MemoryMarshal.Cast` here..

**Comment by PaulusParssinen** [comment]
Assuming the CI failures are unrelated but the fails were about missing log artifact directory, so I'm rolling the CI roulette one more time.

**Comment by CarnaViire** [comment]
@zlatanov JFYI

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 3]
```diff
@@ -649,19 +650,13 @@ private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ReadOnly
...
             else
             {
                 sendBuffer[1] = 127;
-                int length = payload.Length;
-                for (int i = 9; i >= 2; i--)
-                {
-                    sendBuffer[i] = unchecked((byte)length);
-                    length /= 256;
-                }
+                BinaryPrimitives.WriteUInt64BigEndian(sendBuffer.AsSpan(2), (ulong)payload.Length);
```
Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:659]: We do already have a test that would verify it, is that correct?

Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:659]: Also -- I believe there should be a place where it is read "back" -- shouldn't it be changed as well?

Comment (CarnaViire) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:659]: I believe it's here
https://github.com/dotnet/runtime/blob/9be628743487aac580a5e933f08d459dc50c80fb/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L1157-L1173

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 4]
```diff
@@ -1464,7 +1459,7 @@ private void ReleaseSendBuffer()
         }
 
         private static int CombineMaskBytes(Span<byte> buffer, int maskOffset) =>
-            BitConverter.ToInt32(buffer.Slice(maskOffset));
+            BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(maskOffset));
```
Comment (zlatanov) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:1462]: I don't believe this is correct here. The mask is just 4 bytes which were packed as int in the old implementation just for storage.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 2 above)
Comment (zlatanov) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:1493]: I think we should keep the old implementation here, see my comment about the mask packing.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 3 above)
Comment (zlatanov) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:659]: Not explicitly, but the deflate / inflate tests would hit this and verify its correctness.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 2 above)
Comment (PaulusParssinen) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:1493]: Sure, I was on the fence about this one.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 3 above)
Comment (PaulusParssinen) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:659]: The testing coverage is quite dangerously low but a lot of things would have breaked if these were wrong, for example the compression tests as said above😅

**Comment by PaulusParssinen** [comment]
I just realized that `WebSocketValidate.cs` is in shared folder, but.. it's not actually imported by any of the projects in the class and only other place it's "synced" to is azure-relay-dotnet?

https://grep.app/search?q=websocketvalidate&filter[lang][0]=C%23&filter[path][0]=src/&filter[repo][0]=Azure/azure-relay-dotnet

Can we move the class to `System.Net.WebSockets`?

File: src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs [hunk 5]
```diff
@@ -116,22 +138,20 @@ internal static void ValidateCloseStatus(WebSocketCloseStatus closeStatus, strin
...
-            }
-            if (arraySegment.Offset < 0 || arraySegment.Offset > arraySegment.Array.Length)
-            {
-                throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Offset));
-            }
-            if (arraySegment.Count < 0 || arraySegment.Count > (arraySegment.Array.Length - arraySegment.Offset))
-            {
-                throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Count));
-            }
+            string arrayParameterName = parameterName + "." + nameof(arraySegment.Array);
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:145]: This is not an acceptable change. It is now allocating a string regardless of whether the argument is nulll.

File: src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs [hunk 6]
```diff
@@ -37,7 +47,19 @@ internal static partial class WebSocketValidate
...
+            => ThrowIfInvalidState(currentState, isDisposed, s_validSendStates);
+
+        internal static void ThrowIfInvalidReceiveState(WebSocketState currentState, bool isDisposed)
+            => ThrowIfInvalidState(currentState, isDisposed, s_validReceiveStates);
+
+        internal static void ThrowIfInvalidCloseOutputState(WebSocketState currentState, bool isDisposed)
+            => ThrowIfInvalidState(currentState, isDisposed, s_validCloseOutputStates);
+
+        internal static void ThrowIfInvalidCloseState(WebSocketState currentState, bool isDisposed)
+            => ThrowIfInvalidState(currentState, isDisposed, s_validCloseStates);
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:60]: Why is adding all these methods beneficial?

File: src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 5 above)
Comment (PaulusParssinen) [reply src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:145]: Ah right! Thanks for spotting. Will revert this.

File: src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 6 above)
Comment (PaulusParssinen) [reply src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs:60]: ``WebSocketValidate.ThrowIfInvalidState(currentState, isDisposed, WebSocketValidate.ValidCloseStates);`` felt too long winded. I don't feel strongly about these, will revert 👍

File: src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 5 above)
Comment (PaulusParssinen) [reply src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: Custom interpolated string handler short-circuit would be nice in the new throw helpers for this 😅

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: > Not worth it probably.

No. There would still be additional checks performed. In the production path where the argument should _never_ be null, we want the overhead to be nothing more than a single comparison/branch.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketReceiveResult.cs [hunk 7]
```diff
@@ -1,8 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics;
```
Comment (zlatanov) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketReceiveResult.cs:4]: I would abstain from removing unused namespaces from files in this PR as it is not related, and especially when it's the only change in the file.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketException.cs [hunk 8]
```diff
@@ -2,7 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.ComponentModel;
-using System.Diagnostics.CodeAnalysis;
```
Comment (zlatanov) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketException.cs:5]: See my other comment about namespaces.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketCloseStatus.cs [hunk 9]
```diff
@@ -1,8 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics.CodeAnalysis;
```
Comment (zlatanov) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketCloseStatus.cs:4]: See my other comment about namespaces.

**Review by zlatanov — APPROVED** [review]

**Comment by zlatanov** [comment]
LGTM. Thanks!

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks! (and sorry for the delay)

**Pull request merged**